cpp/src/io/parquet/page_string_decode.cu

```diff
@@ -0,0 +1,1125 @@
+
+  // if we're skipping this page anyway, no need to count it
+  if (!is_bounds_pg && !is_page_contained(s, min_row, num_rows)) { return; }
+
+  // find start/end value indices
+  auto const [start_value, end_value] =
+    page_bounds<lvl_buf_size>(s, min_row, num_rows, is_bounds_pg, has_repetition, decoders);
+
+  // need to save num_nulls and num_valids calculated in page_bounds in this page
+  if (t == 0) {
+    pp->num_nulls  = s->page.num_nulls;
+    pp->num_valids = s->page.num_valids;
+  }
+
+  // now process string info in the range [start_value, end_value)
+  // set up for decoding strings...can be either plain or dictionary
+  auto const& col          = s->col;
+  uint8_t const* data      = s->data_start;
+  uint8_t const* const end = s->data_end;
+  uint8_t const* dict_base = nullptr;
+  int dict_size            = 0;
+  size_t str_bytes         = 0;
+
+  switch (pp->encoding) {
+    case Encoding::PLAIN_DICTIONARY:
+    case Encoding::RLE_DICTIONARY:
+      // RLE-packed dictionary indices, first byte indicates index length in bits
+      if (col.str_dict_index) {
+        // String dictionary: use index
+        dict_base = reinterpret_cast<const uint8_t*>(col.str_dict_index);
+        dict_size = col.page_info[0].num_input_values * sizeof(string_index_pair);
+      } else {
+        dict_base = col.page_info[0].page_data;  // dictionary is always stored in the first page
+        dict_size = col.page_info[0].uncompressed_page_size;
+      }
+
+      // FIXME: need to return an error condition...this won't actually do anything
+      if (s->dict_bits > 32 || !dict_base) { CUDF_UNREACHABLE("invalid dictionary bit size"); }
+
+      str_bytes = totalDictEntriesSize(
+        data, dict_base, s->dict_bits, dict_size, (end - data), start_value, end_value);
+      break;
+    case Encoding::PLAIN:
+      dict_size = static_cast<int32_t>(end - data);
+      str_bytes = is_bounds_pg ? totalPlainEntriesSize(data, dict_size, start_value, end_value)
+                               : dict_size - sizeof(int) * (pp->num_input_values - pp->num_nulls);
+      break;
+  }
+
+  if (t == 0) {
+    // TODO check for overflow
+    pp->str_bytes = str_bytes;
+  }
+}
+
+/**
+ * @brief Kernel for computing the string column data stored in the pages
+ *
+ * This function will write the page data and the page data's validity to the
+ * output specified in the page's column chunk.
+ *
+ * This version uses a single warp to do the string copies.
+ *
+ * @param pages List of pages
+ * @param chunks List of column chunks
+ * @param min_row Row index to start reading at
+ * @param num_rows Maximum number of rows to read
+ * @tparam lvl_buf_size Size of the buffer used when decoding repetition and definition levels
+ * @tparam level_t Type used to store decoded repetition and definition levels
+ */
+template <int lvl_buf_size, typename level_t>
+__global__ void __launch_bounds__(decode_block_size) gpuDecodeStringPageData(
```

Comment:
Couple of things about these two functions `gpuDecodeStringPageData` and `gpuDecodeStringPageDataV2`.  As they are right now, these functions will be bottlenecked similar to the non-string path.  1 warp traversing all the level data.  I have a PR almost ready that parallelizes gpuDecodePageData with rle_stream.   

This will undo those wins for string data unless they also get converted to use rle_stream.  But it's a fair amount of work to do this conversion and we now have two additional functions that will need to get it done.  Is there any possibility of somehow merging `gpuDecodeStringPageData` and `gpuDecodeStringPageDataV2` so that they're just a single function with some branching?

Reply:
V2 was written with `rle_stream` in mind.  I envision a loop of:
```
 loop
   decode batch of level data
   decode dict if necessary
   write string data
end loop
```
V2 currently does the first two steps concurrently with 2 warps, and then does step 3 using all 4 warps. I hoping it won't be too much work to modify V2 to work with rle_stream 2.0.

Reply:
I just saw your other reply that the V2 function probably ends up going away.  That reduces the stickiness.

Reply:
Looking at `gpuDecodeStringPageData` I strongly suspect it might be possible to graft the actual decoding work here directly into the new `gpuDecodePageData() `.   The rest of the function structure looks mostly the same and I took the chunk of work that does the actual value copying and separated it into a new function `gpuDecodeValues()`.  It might be possible to just call into this string copying stuff from there. 

It's a big enough change that we probably don't want to hold this PR up on it.  It can be a followup (not a small amount of work though).

Reply:
> V2 was written with `rle_stream` in mind. I envision a loop of:
> 
> ```
>  loop
>    decode batch of level data
>    decode dict if necessary
>    write string data
> end loop
> ```

Yeah this is exactly what we need.   A quick preview of the new gpuDecodePageData

```
while (processed < s->page.num_input_values) {
    int const this_processed = decoders[level_type::DEFINITION].decode_next(t);
    __syncthreads();

    int const this_valid = gpuUpdateValidityOffsetsAndRowIndicesFlat<lvl_buf_size, level_t>(processed + this_processed,
                                                                              s,
                                                                              &shared,
                                                                              def,
                                                                              t);
    __syncthreads();
  
    // if we have dictionary data, decode it now.
    if(s->dict_base){
      dict_stream.decode_next(t, this_valid);
      __syncthreads();
    }    

    // decode the values themselves
    gpuDecodeValues(s, sb, valid, valid + this_valid, t);
    __syncthreads();
  }
```

So if the string copying can be transplanted into gpuDecodeValues (which is basically everything in the last else statement in the existing decode function) then we're all good.

Reply:
Do you have your branch in github yet so I could take a peek? I'm ok with putting this work on hold and seeing how it fits with the new decoder. Plus there's also #13437 in play, and the delta decoding which relies heavily on this PR to avoid 2X the string copies.